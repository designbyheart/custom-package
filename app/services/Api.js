const agencyBaseUrl = 'https://agency.evernym.com/agent'
const callCenterBaseUrl = 'https://callcenter.evernym.com/agent'

export const enroll = device => {
  return fetch(callCenterBaseUrl, {
    method: 'POST',
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(device),
  })
}

export const poll = identifier =>
  fetch(`${agencyBaseUrl}/id/${identifier}/auth`, {
    mode: 'cors',
  })
    .then(res => {
      if (res.status == 200) {
        return res.json()
      } else {
        throw new Error('Bad Request')
      }
    })
    .then(resData => resData.status)
    .catch(error => error)

export const sendAppContext = context => {
  return fetch(`${callCenterBaseUrl}/app-context`, {
    method: 'POST',
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(context),
  })
}

export const sendAuthRequest = ({ identifier, dataBody }) => {
  return fetch(`https://agency.evernym.com/agent/id/${identifier}/auth`, {
    method: 'PUT',
    mode: 'cors',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(dataBody),
  })
}
