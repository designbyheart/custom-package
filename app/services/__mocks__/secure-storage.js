let secureStorage = {
  ConnectMeKeyChain: {},
}

export const setItem = (key, data) => {
  secureStorage['ConnectMeKeyChain'][key] = data
}

export const getItem = key =>
  Promise.resolve(secureStorage['ConnectMeKeyChain'][key])

export const deleteItem = key => delete secureStorage['ConnectMeKeyChain'][key]
