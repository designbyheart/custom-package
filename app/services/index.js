export * from './api'
export * from './keys'
export * from './secure-storage'
export * from './mapper'
export { default as schemaValidator } from './schema-validator'
export * from './error/error-handler'
