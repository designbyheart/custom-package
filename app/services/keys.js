import nacl from 'tweetnacl'
import bs58 from 'bs58'

const decodeUTF8 = s => {
  if (typeof s !== 'string') throw new TypeError('expected string')
  var i, d = unescape(s), b = new Uint8Array(d.length)
  for (i = 0; i < d.length; i++)
    b[i] = d.charCodeAt(i)
  return b
}

export const getKeyPairFromSeed = seed =>
  nacl.sign.keyPair.fromSeed(decodeUTF8(seed))

export const getSignature = (secretKey, message) =>
  nacl.sign.detached(decodeUTF8(message), secretKey)

export const verifySignature = (signature, publicKey, message) =>
  nacl.sign.detached.verify(decodeUTF8(message), signature, publicKey)

export const randomSeed = length => bs58.encode(nacl.randomBytes(length))
