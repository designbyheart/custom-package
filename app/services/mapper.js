import { bubbleSize } from '../common/styles'

export function* generatePlaceholderConnection() {
  let index = 0
  const connections = [
    {
      name: 'evernym',
      image: require('../images/cb_evernym.png'),
      size: bubbleSize.XL,
    },
    {
      name: 'edcu',
      image: require('../images/cb_EDCU.png'),
      size: bubbleSize.XL,
    },
  ]
  while (index >= 0) {
    yield connections[index]
    index++
    if (index > 1) index = 0
  }
}

export const stringToLower = str => str.toLowerCase()

export const connectionMapper = ({
  identifier,
  name = 'suncoast',
  logoUrl,
  size = bubbleSize.XL,
  remoteConnectionId,
}) => ({
  identifier,
  name: stringToLower(name.split(' ')[0]),
  image: logoUrl ? { uri: logoUrl } : require('../images/cbSunCoast.png'),
  size,
  fullName: name,
  remoteConnectionId,
})
