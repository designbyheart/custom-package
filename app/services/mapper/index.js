// @flow

export * from './mapper'
export * from './type-mapper'
