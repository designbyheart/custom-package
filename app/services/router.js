// @flow
import {
  PUSH_NOTIFICATION_SENT_CODE,
  PUSH_NOTIFICATION_TYPE,
  lockEnterPinRoute,
  claimOfferRoute,
  invitationRoute,
} from '../common'
import { CLAIM_OFFER_STATUS } from '../claim-offer/type-claim-offer'

// TODO:KS Add type for flow, before 461 is done, complete types
export default function handlePushNotification(
  props: any,
  notification: any,
  expectedScreen: string,
  isAppLocked?: boolean,
  pendingRedirectAction?: (routeName: string) => void
) {
  if (notification.type === PUSH_NOTIFICATION_TYPE.AUTH) {
    // TODO: change push-notification-sent to push-notification-received
    if (props.route.currentScreen == expectedScreen) {
      props.authenticationRequestReceived({
        offerMsgTitle: notification.authNotifMsgTitle,
        offerMsgText: notification.authNotifMsgText,
        statusCode: PUSH_NOTIFICATION_SENT_CODE,
        logoUrl: notification.logoUrl,
        remoteConnectionId: notification.remoteConnectionId || null,
      })
      if (!isAppLocked) {
        props.navigation.navigate(invitationRoute)
      } else {
        pendingRedirectAction && pendingRedirectAction(invitationRoute)
      }
      props.pushNotificationReceived(null)
    }
  }

  if (
    notification.type === PUSH_NOTIFICATION_TYPE.CLAIM_OFFER &&
    props.route.currentScreen === expectedScreen &&
    expectedScreen !== lockEnterPinRoute
  ) {
    // if we receive claim-offer push notification while we are on lock screen
    // we won't redirect user, because claim offer is supposed to open
    // as a modal pop up, which slides up and then slide down back
    // so, it should open after user is redirected to home screen or connection
    // Other screens will handle redirecting user to claim-offer screen
    // and when claim offer is accepted/ignored or close button is pressed
    // claim offer will slide down and back view will be shown
    // we need to somehow handle all push notification and redirection related
    // stuff in our own navigator which will be built on top of StackNavigator
    props.navigation.navigate(claimOfferRoute)
  }
}
