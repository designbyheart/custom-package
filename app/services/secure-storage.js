// @flow
import RNSensitiveInfo from 'react-native-sensitive-info'
import {
  setWalletItem,
  getWalletItem,
  deleteWalletItem,
} from '../bridge/react-native-cxs/RNCxs'

const storageName = {
  sharedPreferencesName: 'ConnectMeSharedPref',
  keychainService: 'ConnectMeKeyChain',
}

export const setItem = (key: string, data: string) =>
  RNSensitiveInfo.setItem(key, data, storageName)

export const getItem = (key: string) =>
  RNSensitiveInfo.getItem(key, storageName)

export const deleteItem = (key: string) =>
  RNSensitiveInfo.deleteItem(key, storageName)

export const setWalletRecordItem = (key: string, data: string) =>
  setWalletItem(key, data)

export const getWalletRecordItem = (key: string) => getWalletItem(key)

export const deleteWalletRecordItem = (key: string) => deleteWalletItem(key)
