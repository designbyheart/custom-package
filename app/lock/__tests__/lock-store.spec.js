// @flow
import { put, call } from 'redux-saga/effects'
import type { LockStore } from '../type-lock'
import {
  SET_PIN,
  PIN_STORAGE_KEY,
  CHECK_PIN,
  CHECK_PIN_IDLE,
  LOCK_ENABLE,
  SALT_STORAGE_KEY,
} from '../type-lock'
import lockReducer, {
  addPendingRedirection,
  clearPendingRedirect,
  setPin,
  lockEnable,
  checkPin,
  checkPinSuccess,
  checkPinFail,
  setPinAction,
  checkPinStatusIdle,
} from '../lock-store'
import { setItem, getItem } from '../../services/secure-storage'
import { generateSalt, pinHash } from '../pin-hash'

const initialState: LockStore = {
  pendingRedirection: null,
  checkPinStatus: CHECK_PIN_IDLE,
  // we are assuming that app will be locked by default
  // and it will be unlocked either when user set security first time
  // or user unlock the app every time user opens the app
  // this property needs to be set accordingly
  isAppLocked: true,
  showDevMode: false,
  isLockEnabled: false,
  isTouchIdEnabled: false,
}

describe('LockStore', () => {
  it('add pending redirection change state correctly', () => {
    const expectedState = lockReducer(
      initialState,
      addPendingRedirection([{ routeName: 'home' }])
    )
    expect(expectedState).toMatchSnapshot()
  })

  it('clear pending redirect if clear action is raised', () => {
    const pendingState = lockReducer(
      initialState,
      addPendingRedirection([{ routeName: 'home' }])
    )
    const expectedState = lockReducer(pendingState, clearPendingRedirect())
    expect(expectedState).toMatchSnapshot()
  })

  it('set pin should work fine', () => {
    const pin = '123456'
    const salt = 'salt'
    const gen = setPin(setPinAction(pin))
    expect(gen.next().value).toEqual(call(generateSalt))
    const hashedPin: any = gen.next(salt).value
    expect(hashedPin).toEqual(call(pinHash, pin, salt))

    const pinStorage: any = gen.next(pin).value
    expect(pinStorage).toEqual(call(setItem, PIN_STORAGE_KEY, pin))

    const saltStorage: any = gen.next(salt).value
    expect(saltStorage).toEqual(call(setItem, SALT_STORAGE_KEY, salt))

    const lockEnableAction: any = gen.next(pin).value
    expect(lockEnableAction['PUT'].action).toEqual(
      expect.objectContaining({
        type: LOCK_ENABLE,
      })
    )

    expect(gen.next().done).toBe(true)
  })

  it('check pin flow should work if correct pin is passed', () => {
    const pin = '123456'
    const salt = 'salt'
    const gen = checkPin({ type: CHECK_PIN, pin })
    expect(gen.next().value).toEqual(call(getItem, SALT_STORAGE_KEY))
    expect(gen.next(salt).value).toEqual(call(pinHash, pin, salt))
    expect(gen.next(pin).value).toEqual(call(getItem, PIN_STORAGE_KEY))
    expect(gen.next(pin).value).toEqual(put(checkPinSuccess()))
    expect(gen.next().done).toBe(true)
  })

  it('check pin flow should fail if incorrect pin is passed', () => {
    const pin = '123456'
    const salt = 'salt'
    const wrongPin = '123444'
    const gen = checkPin({ type: CHECK_PIN, pin })
    expect(gen.next().value).toEqual(call(getItem, SALT_STORAGE_KEY))
    expect(gen.next(salt).value).toEqual(call(pinHash, pin, salt))
    expect(gen.next(pin).value).toEqual(call(getItem, PIN_STORAGE_KEY))
    expect(gen.next(wrongPin).value).toEqual(put(checkPinFail()))
    expect(gen.next().done).toBe(true)
  })

  it('should set checkPinStatus to idle', () => {
    const failedState = lockReducer(initialState, checkPinFail())
    expect(failedState).toMatchSnapshot()
    const expectedState = lockReducer(failedState, checkPinStatusIdle())
    expect(expectedState).toMatchSnapshot()
  })
})
