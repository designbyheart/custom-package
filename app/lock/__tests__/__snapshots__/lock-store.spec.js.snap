// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LockStore add pending redirection change state correctly 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "isAppLocked": true,
  "isLockEnabled": false,
  "pendingRedirection": Array [
    Object {
      "routeName": "home",
    },
  ],
}
`;

exports[`LockStore clear pending redirect if clear action is raised 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "isAppLocked": true,
  "isLockEnabled": false,
  "pendingRedirection": null,
}
`;

exports[`LockStore should set checkPinStatus to idle 1`] = `
Object {
  "checkPinStatus": "CHECK_PIN_FAIL",
  "isAppLocked": true,
  "isLockEnabled": false,
  "pendingRedirection": null,
}
`;

exports[`LockStore should set checkPinStatus to idle 2`] = `
Object {
  "checkPinStatus": "CHECK_PIN_IDLE",
  "isAppLocked": true,
  "isLockEnabled": false,
  "pendingRedirection": null,
}
`;
